source/common/stats/symbol_table_impl.cc

```diff
@@ -126,69 +126,109 @@ SymbolVec SymbolTableImpl::Encoding::decodeSymbols(const SymbolTable::Storage ar
-  while (size > 0) {
-    if (*array == LiteralStringIndicator) {
+// Iterator for decoding StatName tokens. This is not an STL-compatible
+// iterator; it is used only locally. Using this rather than decodeSymbols makes
+// it easier to early-exit out of a comparison once we know the answer. For
+// example, if you are ordering "a.b.c.d" against "a.x.y.z" you know the order
+// when you get to the "b" vs "x" and do not need to decode "c" vs "y".
+//
+// This is also helpful for prefix-matching.
+//
+// The array-version of decodeSymbols is still a good approach for converting a
+// StatName to a string as the intermediate array of string_view is needed to
+// allocate the right size for the joined string.
+class SymbolTableImpl::Encoding::TokenIter {
+public:
+  // The type of token reached.
+  enum class Type { StringView, Symbol, End };
+
+  TokenIter(const SymbolTable::Storage array, size_t size) : array_(array), size_(size) {}
+
+  /**
+   * Parses the next token. The token can then be retrieved by calling
+   * stringView() or symbol().
+   * @return The type of token, Type::End if we have reached the end of the StatName.
+   */
+  Type next() {
+    if (size_ == 0) {
+      return Type::End;
+    }
+    if (*array_ == LiteralStringIndicator) {
       // To avoid scanning memory to find the literal size during decode, we
       // var-length encode the size of the literal string prior to the data.
-      ASSERT(size > 1);
-      ++array;
-      --size;
-      std::pair<uint64_t, size_t> length_consumed = decodeNumber(array);
+      ASSERT(size_ > 1);
+      ++array_;
+      --size_;
+      std::pair<uint64_t, size_t> length_consumed = decodeNumber(array_);
       uint64_t length = length_consumed.first;
-      array += length_consumed.second;
-      size -= length_consumed.second;
-      ASSERT(size >= length);
-      string_view_token_fn(absl::string_view(reinterpret_cast<const char*>(array), length));
-      size -= length;
-      array += length;
-    } else {
-      std::pair<uint64_t, size_t> symbol_consumed = decodeNumber(array);
-      symbol_token_fn(symbol_consumed.first);
-      size -= symbol_consumed.second;
-      array += symbol_consumed.second;
+      array_ += length_consumed.second;
+      size_ -= length_consumed.second;
+      ASSERT(size_ >= length);
+      string_view_ = absl::string_view(reinterpret_cast<const char*>(array_), length);
+      size_ -= length;
+      array_ += length;
+      return Type::StringView;
     }
+    std::pair<uint64_t, size_t> symbol_consumed = decodeNumber(array_);
+    symbol_ = symbol_consumed.first;
+    size_ -= symbol_consumed.second;
+    array_ += symbol_consumed.second;
+    return Type::Symbol;
   }
-}
 
-bool StatName::startsWith(StatName symbolic_prefix) const {
-  bool ret = true;
-  std::vector<Symbol> prefix_symbols;
+  /** @return the current string_view -- only valid to call if next()==Type::StringView */
+  absl::string_view stringView() const { return string_view_; }
```

Comment:
This should either return optional, or should assert that next() returned Type::StringView

Reply:
I think I'd do that if this were an exposed interface, but this is internal to this file, and is performance critical, so I'd prefer not do do either of the above. I will note that in a comment though.

Reply:
I am OK doing this with  an `#Ifndef NDEBUG` guarded extra member var to remember the previous return from next(); WDYT?

Reply:
actually this wasn't that bad, so I went ahead and did it.

Reply:
That might get gnarly given the multiple returns and if else's. Might be simplified if the returns can be wrapped in a macro which sets the value on the variable when available.

If so, it might be a good compromise.